components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java

```diff
@@ -2522,6 +2536,83 @@ public String createApplication(ServiceProvider application, String tenantDomain
+     *
+     * @param applicationModel ApplicationModelDTO containing the app information
+     * @param tenantDomain     Tenant domain.
+     * @param username         Username.
+     * @return Application ID.
+     * @throws IdentityApplicationManagementException Identity Application Management Exception.
+     */
+    public String createApplication(ApplicationDTO applicationModel, String tenantDomain, String username)
+            throws IdentityApplicationManagementException {
+
+        List<InboundAuthenticationRequestConfig> addedInbounds = new ArrayList<>();
+        try {
+            ServiceProvider application = applicationModel.getServiceProvider();
+             addedInbounds = addInboundAuthenticationProtocolsToApplication(
+                    application, applicationModel.getInboundProtocolConfigurationDto());
+            
+            return createApplication(application, tenantDomain, username);
+        } catch (Exception e) {
+            /*
+             * The current implementation of the application creation process is not atomic. Therefore, if an Exception
+             * occurs, there is a chance that the database gets updated partially. Hence, we need to rollback the
+             * created inbounds if available
+             * For more information read https://github.com/wso2/product-is/issues/12579.
+             */
+            rollbackInbounds(addedInbounds);
+            throw ApplicationMgtUtil.handleException("Server encountered an unexpected error when " +
+                    "creating the application.", e);
+        }
+    }
+
+    private List<InboundAuthenticationRequestConfig> addInboundAuthenticationProtocolsToApplication(
+            ServiceProvider application, InboundProtocolsDTO inboundProtocolsModel)
+            throws IdentityApplicationManagementException {
+
+        if (inboundProtocolsModel == null) {
+            return Collections.emptyList();
+        }
+        List<InboundAuthenticationRequestConfig> addedInbounds = new ArrayList<>();
+        try {
+            // This handles the inbound protocol creation with the help of protocol handlers.
+            List<ApplicationInboundAuthConfigHandler> applicationInboundAuthConfigHandlerList =
+                    ApplicationManagementServiceComponentHolder.getInstance().getApplicationInboundAuthConfigHandler();
+            for (ApplicationInboundAuthConfigHandler applicationInboundAuthConfigHandler :
+                    applicationInboundAuthConfigHandlerList) {
+                if (applicationInboundAuthConfigHandler.canHandle(inboundProtocolsModel)) {
+                    addedInbounds.add(applicationInboundAuthConfigHandler.handleConfigCreation(application,
+                            inboundProtocolsModel));
+                }
+            }
+        } catch (IdentityApplicationManagementClientException e) {
+            // This is to void converting client exception in to a server exception.
```

Comment:
Shall we add a comment on why roll back is not required for the client exceptions

Reply:
Fixed with 2eb7f05bc4f9dd13bebb13286d5b205745ffdad9